core/src/main/java/org/apache/druid/data/input/impl/MapInputRowParser.java

```diff
@@ -36,20 +36,22 @@
 public class MapInputRowParser implements InputRowParser<Map<String, Object>>
 {
   private final ParseSpec parseSpec;
+  private final List<String> dimensions;
 
   @JsonCreator
   public MapInputRowParser(
       @JsonProperty("parseSpec") ParseSpec parseSpec
   )
   {
     this.parseSpec = parseSpec;
+    this.dimensions = parseSpec.getDimensionsSpec().getDimensionNames();
   }
 
   @Override
   public List<InputRow> parseBatch(Map<String, Object> theMap)
   {
-    final List<String> dimensions = parseSpec.getDimensionsSpec().hasCustomDimensions()
-                                    ? parseSpec.getDimensionsSpec().getDimensionNames()
+    final List<String> dimensions = !this.dimensions.isEmpty()
```

Comment:
Please use if-else instead of multiline ternary. In other places too.

Reply:
I see this same basic pattern in a lot of `InputRowParser` implementations too, maybe could be shared method?

Reply:
@clintropolis I think so, but am not sure where to place this shared method. Add a default method in interface InputRowParser like below?
```java
default List<String> getDimensions(ParseSpec parseSpec, List<String> dimensions, Map<String, Object> row) 
  {
    final List<String> rv;
    if (!dimensions.isEmpty()) {
      rv = dimensions;
    } else {
      rv = Lists.newArrayList(
          Sets.difference(row.keySet(), parseSpec.getDimensionsSpec().getDimensionExclusions())
      );
    }
    return rv;
  }
```

Reply:
I think there shouldn't be a method. Because adding it feels like net negative in terms of the overall complexity.

Reply:
@leventov ok, fine. Have updated.